src/hash_traits.jl

```diff
@@ -0,0 +1,461 @@
+    check_hash_method(x, transform, context)
+    return hash_value!(tx, hash_state, context, hash_trait(transform, tx))
+end
+
+# There are two cases where we want to hash types:
+#
+#   1. when we are hashing the type of an object we're hashing (`TypeHashContext`)
+#   2. when a value we're hashing is itself a type (`TypeAsValueContext`)
+#
+# These are handled as separate contexts because the kind of value we want to generate from
+# the type may differ. By default only the structure of types matters when hashing an
+# objects type, e.g. when we hash a StructTypes.DataType we hash that it is a data type, the
+# field names and we hash each individual element type (as per its rules) but we do not hash
+# the name of the type. When a type is hashed as a value, its actual name also matters.
+
+#####
+##### Type Hashes
+#####
+
+function transformer(::Type{T}, context::TypeHashContext) where {T<:Type}
+    return Transformer(T -> (type_hash_name(T, StructType_(T), context),
+                             type_structure(T, StructType_(T), context)))
+end
+@inline StructType_(T) = StructType(T)
+StructType_(::Type{Union{}}) = StructTypes.NoStructType()
+
+"""
+    type_hash_name(::Type{T}, trait, [context])
+
+The name that is hashed for type `T` when hashing the type of a given value. This defaults
+to `stable_type_name(trait)`. Users of `StableHashTraits` can implement a method that accepts two
+(`T` and `trait`) or three arguments (`T`, `trait`, `context`). The trait is one of the
+`StructTypes` traits and `context` is the hash context (as per the second argument to
+`stable_hash`)
+"""
+function type_hash_name(::Type{T}, trait, context) where {T}
+    return type_hash_name(T, trait, parent_context(context))
+end
+type_hash_name(::Type{T}, trait, ::Nothing) where {T} = type_hash_name(T, trait)
+type_hash_name(::Type{T}, trait) where {T} = qualified_name_(trait)
+
+"""
+    type_structure(::Type{T}, trait, [context])
+
+Get the types and symbols that represent the structure of `T`. This should return the
+relevant, contained types from `T` that should be hashed. The trait is one of the
+`StructTypes` traits and `context` is the hash context (as per the second argument to
+`stable_hash`). The second must be included as part of a method definition, but `context` is
+optional.
+
+For example, this is the definition for `ArrayType` objects.
+
+```julia
+function type_structure(::Type{T}, ::StructTypes.ArrayType) where {T}
+    return eltype(T)
+end
+```
+
+You need to override this method when a type parameter of `T` is important to the type's
+hash but is not included in `fieldtypes(T)`. For example:
+
+```julia
+struct MyType{T,F}
+    obj::T
+end
+StableHashTratis.type_structure(::Type{T}, ::StructTypes.DataType) where {F,T<:MyType{<:Any,F}}
+    return fieldnames(T), fieldtypes(T), F
+end
+```
+
+Without this definition `MyType{Int,:foo}(1)` would hash to the same value as
+`MyType{Int,:bar}(2)`.
+
+!!! warning "Hash all expected type structure!"
```

Comment:
I realized today that there is a better way to implement `type_structure` that provides more guardrails and means this warning doesn't have to be added to the documentation.

I can just have internals call `type_structure_helper` which always adds the required structure, and `type_structure` can be the public facing method that is expected to return any custom structure the user wants to be added.